py_ecc/bn128_curve.py

```diff
@@ -0,0 +1,180 @@
+from typing import (
+    cast,
+)
+
+from py_ecc.BaseCurve import (
+    BaseCurve,
+    BaseOptimizedCurve,
+)
+
+from py_ecc.curve_properties import (
+    curve_properties,
+    optimized_curve_properties,
+)
+
+from py_ecc.field_elements import (
+    FQP,
+    FQ12,
+)
+
+from py_ecc.field_properties import (
+    field_properties,
+)
+
+from py_ecc.optimized_field_elements import (
+    FQP as optimized_FQP,
+    FQ12 as optimized_FQ12,
+)
+
+from py_ecc.typing import (
+    Field,
+    Optimized_Point3D,
+    Point2D,
+)
+
+
+class BN128_Curve(BaseCurve):
+    curve_name = "bn128"
+    curve_order = curve_properties[curve_name]["curve_order"]
```

Comment:
I am finding this structure awkward and somewhat hard to read.  It seems like this would be significantly simpler as something like.

```python
# py_ecc/field_properties.py
bn128_props = typing.NamedTuple(  # using 3.5 syntax because I think we have to support it
    'bn128_props',
    (
        ('b', int),
        ...
    )
)

# py_ecc/bn128_curve.py
from py_ecc.field_properties import bn128_props

class BN128_Curve(BaseCurve):
    field_modulus = bn128_props.b
    ...
```

Benefits I see:

- Removes the string based field lookups: `thing.prop` instead of `thing["prop"]`
- Removes the nested lookups: `thing.prop` instead of `outer_things[thing_name]["prop"]`